### PR TITLE
[lodash] Support WeakMap's for memoize.Cache

### PR DESCRIPTION
**Pull request by wyattjoh**
This is reintroducing the attempted PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24487 by @cspotcode that addresses the issues highlighted in https://github.com/lodash/lodash/issues/3705 by marking the `.clear()` method as optional, and changing the signatures for the remaining methods to support `any` instead of `string`.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/lodash/lodash/issues/3705
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by wyattjoh** [comment]
Seems this test overwhelmed the RAM on the TravisCI box 😢

**Comment by cspotcode** [comment]
@wyattjoh Thank you!  And sorry that I forgot about my original PR.  :(

**Comment by wyattjoh** [comment]
Hey @bczengel @chrootsu @stepancar @aj-r @Ailrun @e-cloud @thorn0 @jtmthf @DomiR, it looks like the Travis build is failing from an unrelated error, anything we can do to get this resolved/merged?

**Comment by wyattjoh** [comment]
Thanks for the suggestion @thorn0, I've closed + reopened the PR as well as rebased onto master. Fingers crossed CI doesn't die again!

**Comment by wyattjoh** [comment]
The error that CI is throwing is specifically because Travis is running out of memory:

```
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```

I doubt it's related to my changes.

**Comment by thorn0** [comment]
I checked that the same CI error doesn't happen in other PRs, so it's the proposed changes that caused it. I see two directions to understand it better:

1. Revert the changes line by line to see when the error disappears.
2. Gradually increase the minimal required TS version to see the same. Might be a memory-related bug in some specific TS version.

**Comment by wyattjoh** [comment]
So after following the steps outlined here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24052#pullrequestreview-101008220, I was able to get it to print the actual error:

```
ERROR: 3490:5  expect  Compile error in typescript@2.9 but not in typescript@3.0.
Fix with a comment '// TypeScript Version: 3.0' just under the header.
Type 'WeakMap<object, any>' is not assignable to type 'MapCache'.
  Types of property 'set' are incompatible.
    Type '(key: object, value: any) => WeakMap<object, any>' is not assignable to type '(key: string, value: any) => Dictionary<any>'.
      Types of parameters 'key' and 'key' are incompatible.
        Type 'string' is not assignable to type 'object'.
```

So to address that issue, I had to change the signature of `set`: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30035/commits/5e8e4d298b0ea95035883a823b95cc6ddece065d.

File: types/lodash/common/common.d.ts [hunk 1]
```diff
@@ -149,7 +149,7 @@ declare module "../index" {
          * @param value The value to cache.
          * @return Returns the cache object.
          */
-        set(key: string, value: any): Dictionary<any>;
+        set(key: any, value: any): Dictionary<any>;
```
Comment (thorn0) [inline types/lodash/common/common.d.ts]: The return type should be `this`.

File: types/lodash/common/common.d.ts [hunk 2]
```diff
@@ -127,34 +127,34 @@ declare module "../index" {
...
+        has(key: any): boolean;
 
         /**
          * Sets `value` to `key` of the cache.
          * @param key The key of the value to cache.
          * @param value The value to cache.
          * @return Returns the cache object.
          */
-        set(key: string, value: any): Dictionary<any>;
+        set(key: any, value: any): Dictionary<any>;
```
Comment (thorn0) [inline types/lodash/common/common.d.ts]: The return type should be `this`.

**Review by thorn0 — CHANGES_REQUESTED** [review]

**Review by thorn0 — APPROVED** [review]

**Pull request merged**